crates/bevy_core_pipeline/src/bloom/bloom.wgsl

```diff
@@ -1,138 +1,137 @@
-    let m = textureSample(original, original_sampler, uv + vec2<f32>(1.0, 1.0) * scale);
-
-    let div = (1.0 / 4.0) * vec2<f32>(0.5, 0.125);
-
-    var o: vec4<f32> = (d + e + i + j) * div.x;
-    o = o + (a + b + g + f) * div.y;
-    o = o + (b + c + h + g) * div.y;
-    o = o + (f + g + l + k) * div.y;
-    o = o + (g + h + m + l) * div.y;
-
-    return o;
+// "Next Generation Post Processing in Call of Duty: Advanced Warfare" slide 100
+fn sample_input_13_tap(uv: vec2<f32>) -> vec3<f32> {
+    let texel_size = 1.0 / vec2<f32>(textureDimensions(input_texture));
+    let x = texel_size.x;
+    let y = texel_size.y;
+
+    let a = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y + 2.0 * y)).rgb;
+    let b = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y + 2.0 * y)).rgb;
+    let c = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y + 2.0 * y)).rgb;
+
+    let d = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y)).rgb;
+    let e = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y)).rgb;
+    let f = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y)).rgb;
+
+    let g = textureSample(input_texture, s, vec2<f32>(uv.x - 2.0 * x, uv.y - 2.0 * y)).rgb;
+    let h = textureSample(input_texture, s, vec2<f32>(uv.x, uv.y - 2.0 * y)).rgb;
+    let i = textureSample(input_texture, s, vec2<f32>(uv.x + 2.0 * x, uv.y - 2.0 * y)).rgb;
+
+    let j = textureSample(input_texture, s, vec2<f32>(uv.x - x, uv.y + y)).rgb;
+    let k = textureSample(input_texture, s, vec2<f32>(uv.x + x, uv.y + y)).rgb;
+    let l = textureSample(input_texture, s, vec2<f32>(uv.x - x, uv.y - y)).rgb;
+    let m = textureSample(input_texture, s, vec2<f32>(uv.x + x, uv.y - y)).rgb;
+
+#ifdef FIRST_DOWNSAMPLE
+    // "Next Generation Post Processing in Call of Duty: Advanced Warfare" slide 112
+    var group0 = (a + b + d + e) * (0.125f / 4.0f);
+    var group1 = (b + c + e + f) * (0.125f / 4.0f);
+    var group2 = (d + e + g + h) * (0.125f / 4.0f);
+    var group3 = (e + f + h + i) * (0.125f / 4.0f);
+    var group4 = (j + k + l + m) * (0.5f / 4.0f);
+    group0 *= karis_average(group0);
+    group1 *= karis_average(group1);
+    group2 *= karis_average(group2);
+    group3 *= karis_average(group3);
+    group4 *= karis_average(group4);
+    return group0 + group1 + group2 + group3 + group4;
+#else
+    var sample = e * 0.125;
+    sample += (a + c + g + i) * 0.03125;
+    sample += (b + d + f + h) * 0.0625;
+    sample += (j + k + l + m) * 0.125;
+    return sample;
+#endif
 }
 
-// Samples original using a 3x3 tent filter.
-//
-// NOTE: Use a 2x2 filter for better perf, but 3x3 looks better.
-fn sample_original_3x3_tent(uv: vec2<f32>, scale: vec2<f32>) -> vec4<f32> {
-    let d = vec4<f32>(1.0, 1.0, -1.0, 0.0);
+// "Next Generation Post Processing in Call of Duty: Advanced Warfare" slide 109
+fn sample_input_3x3_tent(uv: vec2<f32>) -> vec3<f32> {
+    let x = 0.004;
+    let y = 0.004;
```

Comment:
Where does this 0.004 come from?

Reply:
I believe this was chosen empirically. It corresponds to the "radius" parameter of this slide. @StarLederer can you confirm if my understanding is correct? 

![image](https://user-images.githubusercontent.com/47158642/218291039-8a3aa40d-859a-4764-a302-a4822ae5a168.png)

Reply:
Yup, it is the filter 'radius' indeed. Whether empirical or otherwise, I think we should document where it comes from for future readers. :)

Reply:
I don't have much space on my computer so I can't really view the source material at the moment. I do remember that there were in fact a lot of precomputed constants and 0.004 is likely one of them. I never really questioned the source material because it is a popular technique.

Reply:
I see 0.005 mentioned in the comments of the learnopengl article, so somewhere around there is I guess an empirically good value.

Reply:
Clarity: please add a comment that this was tweaked based on 0.005 in the learnopengl article. I see in the article and comments the author recommended 0.005 and someone else said they had to use 0.003 to address some artifacts.